

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pymbt.sequence_generation.orthogonal_sequences &mdash; pymbt 0.1a documentation</title>
    
    <link rel="stylesheet" href="../../../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '0.1a',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <link rel="top" title="pymbt 0.1a documentation" href="../../../index.html" />
    <link rel="up" title="Module code" href="../../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../../../index.html">pymbt 0.1a documentation</a> &raquo;</li>
          <li><a href="../../index.html" accesskey="U">Module code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <h1>Source code for pymbt.sequence_generation.orthogonal_sequences</h1><div class="highlight"><pre>
<span class="sd">&#39;&#39;&#39;Generates a set of orthogonal DNA sequences for a given protein sequence</span>
<span class="sd">   Currently limited by the size of the protein sequence - ~10 residues max</span>
<span class="sd">   To increase size, need to optimize search method and take steps to reduce</span>
<span class="sd">   memory usage (write to files and iterate/queue)&#39;&#39;&#39;</span>

<span class="c"># TODO: Many parts of the while loop are repetitive</span>

<span class="kn">import</span> <span class="nn">csv</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">from</span> <span class="nn">datetime</span> <span class="kn">import</span> <span class="n">datetime</span>
<span class="kn">from</span> <span class="nn">itertools</span> <span class="kn">import</span> <span class="n">combinations</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">socket</span>
<span class="kn">import</span> <span class="nn">time</span>

<span class="kn">from</span> <span class="nn">pymbt.nupack</span> <span class="kn">import</span> <span class="n">nupack_multiprocessing</span>
<span class="kn">from</span> <span class="nn">pymbt.nupack</span> <span class="kn">import</span> <span class="n">Nupack</span>
<span class="kn">from</span> <span class="nn">pymbt.sequence_generation</span> <span class="kn">import</span> <span class="n">random_codons</span>
<span class="kn">from</span> <span class="nn">pymbt.sequence_generation</span> <span class="kn">import</span> <span class="n">weighted_codons</span>
<span class="kn">from</span> <span class="nn">pymbt.sequence_manipulation</span> <span class="kn">import</span> <span class="n">reverse_complement</span> <span class="k">as</span> <span class="n">r_c</span>
<span class="kn">from</span> <span class="nn">pymbt.sequence_manipulation</span> <span class="kn">import</span> <span class="n">check_alphabet</span>


<div class="viewcode-block" id="OrthoSeq"><a class="viewcode-back" href="../../../pymbt.sequence_generation.html#pymbt.sequence_generation.orthogonal_sequences.OrthoSeq">[docs]</a><span class="k">class</span> <span class="nc">OrthoSeq</span><span class="p">:</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                 <span class="n">prot_seq</span><span class="p">,</span>
                 <span class="n">T</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
                 <span class="n">min_score</span><span class="o">=</span><span class="mf">0.95</span><span class="p">,</span>
                 <span class="n">n_attempt</span><span class="o">=</span><span class="mf">1e5</span><span class="p">):</span>
        <span class="c"># Make sure sequence only includes amino acids</span>
        <span class="n">check_alphabet</span><span class="p">(</span><span class="n">prot_seq</span><span class="p">,</span> <span class="n">material</span><span class="o">=</span><span class="s">&#39;pep&#39;</span><span class="p">)</span>
        <span class="c"># Input sequence - a peptide</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">prot_seq</span> <span class="o">=</span> <span class="n">prot_seq</span>
        <span class="c"># Temp at which to evaluate all reactions</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">T</span> <span class="o">=</span> <span class="n">T</span>
        <span class="c"># Primary scoring metric - minimum average free monomer concentrations</span>
        <span class="c"># - 0 would be fully-bound, 1 would be nothing bound</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">min_score</span> <span class="o">=</span> <span class="n">min_score</span>
        <span class="c"># Secondary scoring metric - stops the simulation after n</span>
        <span class="c"># attempts that fail to increase the primary score</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n_attempt</span> <span class="o">=</span> <span class="n">n_attempt</span>

<div class="viewcode-block" id="OrthoSeq.start"><a class="viewcode-back" href="../../../pymbt.sequence_generation.html#pymbt.sequence_generation.orthogonal_sequences.OrthoSeq.start">[docs]</a>    <span class="k">def</span> <span class="nf">start</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
              <span class="n">n</span><span class="p">,</span>
              <span class="n">m</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
              <span class="n">resume</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
              <span class="n">wholemat</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
              <span class="n">report</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
              <span class="n">weighted</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
              <span class="n">freq_threshold</span><span class="o">=</span><span class="mf">0.5</span><span class="p">):</span>

        <span class="sd">&#39;&#39;&#39;n: is the final number of orthogonal sequences to output.</span>
<span class="sd">           m: is the number of new sequences to try out on each iteration.</span>
<span class="sd">           resume: specifies a path to a dir from a previous run, and will</span>
<span class="sd">           cause start() to pick up where it left off.</span>
<span class="sd">           wholemat: if True, m worst oligos are thrown out on each iteration.</span>
<span class="sd">                     if False, single best is kept. There is a difference</span>
<span class="sd">                     but it should be better documented.</span>
<span class="sd">           weighted: boolean for whether codons should be weighted.</span>
<span class="sd">           freq_threshold: relative frequency threshold below which codons</span>
<span class="sd">                     should not be used.</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="c"># Concentrations at which simulations are done</span>
        <span class="n">conc</span> <span class="o">=</span> <span class="mf">5e-7</span>
        <span class="c"># Arguments for _gen_oligo sequence generation</span>
        <span class="n">oligo_args</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">prot_seq</span><span class="p">,</span>
                      <span class="bp">self</span><span class="o">.</span><span class="n">min_score</span><span class="p">,</span>
                      <span class="n">freq_threshold</span><span class="p">,</span>
                      <span class="n">weighted</span><span class="p">,</span>
                      <span class="n">conc</span><span class="p">)</span>
        <span class="c"># Generate n oligos that meet score threshold</span>
        <span class="c"># If there&#39;s a list of oligos to resume improving, start using it</span>
        <span class="k">if</span> <span class="n">resume</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">resume</span><span class="p">)</span> <span class="o">!=</span> <span class="n">n</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s">&#39;Number of sequences to resume doesn</span><span class="se">\&#39;</span><span class="s">t match</span><span class="se">\</span>
<span class="s">                                 setting of </span><span class="se">\&#39;</span><span class="s">n</span><span class="se">\&#39;</span><span class="s">.&#39;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">resume</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">str</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\&#39;</span><span class="s">resume</span><span class="se">\&#39;</span><span class="s"> input isn</span><span class="se">\&#39;</span><span class="s">t strings!&#39;</span><span class="p">)</span>
            <span class="n">oligo_list</span> <span class="o">=</span> <span class="n">resume</span>
        <span class="c"># If not, generate initial set of oligos</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">oligo_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">_gen_oligo</span><span class="p">(</span><span class="o">*</span><span class="n">oligo_args</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">)]</span>
            <span class="n">oligo_list</span> <span class="o">=</span> <span class="n">remove_redundant</span><span class="p">(</span><span class="n">oligo_list</span><span class="p">,</span> <span class="n">oligo_args</span><span class="p">)</span>
        <span class="n">time_start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
        <span class="n">loop_count</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">n_attempted</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">last_min_score</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="c"># central loop of the program - keeps running until either the</span>
        <span class="c"># primary score is met or the score fails to increase for n attempts</span>
        <span class="k">while</span> <span class="bp">True</span><span class="p">:</span>
            <span class="n">loop_count</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="c"># Generate new oligo(s) to test in combination with the rest</span>
            <span class="n">new_oligos</span> <span class="o">=</span> <span class="p">[</span><span class="n">_gen_oligo</span><span class="p">(</span><span class="o">*</span><span class="n">oligo_args</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">m</span><span class="p">)]</span>
            <span class="n">oligo_list</span> <span class="o">+=</span> <span class="n">new_oligos</span>
            <span class="c"># If two or more oligos are the same, replace with unique ones</span>
            <span class="n">oligo_list</span> <span class="o">=</span> <span class="n">remove_redundant</span><span class="p">(</span><span class="n">oligo_list</span><span class="p">,</span> <span class="n">oligo_args</span><span class="p">)</span>
            <span class="c"># Generate all combinations - (m+n) choose n</span>
            <span class="n">combos</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">combinations</span><span class="p">(</span><span class="n">oligo_list</span><span class="p">,</span> <span class="n">n</span><span class="p">)]</span>

            <span class="c"># For every combination of length m, generate m new lists where</span>
            <span class="c"># all but one of the oligos is reverse-complemented.</span>
            <span class="c"># reversed_combos is a list of lists of sequences, each entry</span>
            <span class="c"># compatible with the Nupack module&#39;s complexes method</span>
            <span class="n">reversed_combos</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">combo</span> <span class="ow">in</span> <span class="n">combos</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">combo</span><span class="p">:</span>
                    <span class="n">newlist</span> <span class="o">=</span> <span class="p">[</span><span class="n">seq</span><span class="p">]</span>
                    <span class="n">newlist</span> <span class="o">+=</span> <span class="p">[</span><span class="n">r_c</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">combo</span> <span class="k">if</span> <span class="n">x</span> <span class="o">!=</span> <span class="n">seq</span><span class="p">]</span>
                    <span class="n">reversed_combos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">newlist</span><span class="p">)</span>

            <span class="c"># a list of Nupack&#39;s &#39;complexes&#39; that&#39;s m * (m + n) long</span>
            <span class="n">np_concs_raw</span> <span class="o">=</span> <span class="n">nupack_multiprocessing</span><span class="p">(</span><span class="n">reversed_combos</span><span class="p">,</span>
                                                  <span class="s">&#39;dna&#39;</span><span class="p">,</span>
                                                  <span class="s">&#39;concentrations&#39;</span><span class="p">,</span>
                                                  <span class="p">{</span><span class="s">&#39;max_complexes&#39;</span><span class="p">:</span> <span class="mi">2</span><span class="p">})</span>
            <span class="c"># just the concentrations list from each entry</span>
            <span class="n">np_concs</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;concentration&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">np_concs_raw</span><span class="p">]</span>
            <span class="c"># just the types list from each entry</span>
            <span class="n">np_types</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;types&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">np_concs_raw</span><span class="p">]</span>
            <span class="c"># the monomer concentrations from each entry</span>
            <span class="n">mon_concs</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">np_types</span><span class="p">):</span>
                <span class="c"># if sum of the list is 1, that means it&#39;s a monomer conc</span>
                <span class="c"># (e.g. [1, 0, 0, 0])</span>
                <span class="n">temp_list</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
                    <span class="k">if</span> <span class="nb">sum</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="n">temp_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np_concs</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">])</span>
                <span class="n">mon_concs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">temp_list</span><span class="p">)</span>

            <span class="c"># Group by combo</span>
            <span class="n">grouped_mon_concs</span> <span class="o">=</span> <span class="n">unflatten</span><span class="p">(</span><span class="n">mon_concs</span><span class="p">,</span> <span class="n">m</span><span class="p">)</span>

            <span class="c"># Find the worst binder within each of the m</span>
            <span class="c"># rev-comp variations per combo</span>
            <span class="n">grouped_mins</span> <span class="o">=</span> <span class="p">[[</span><span class="nb">min</span><span class="p">(</span><span class="n">y</span><span class="p">)</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">grouped_mon_concs</span><span class="p">]</span>

            <span class="c"># Find the worst binder for each combination</span>
            <span class="n">mon_conc_mins</span> <span class="o">=</span> <span class="p">[</span><span class="nb">min</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">grouped_mins</span><span class="p">]</span>
            <span class="n">best_index</span> <span class="o">=</span> <span class="n">mon_conc_mins</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">mon_conc_mins</span><span class="p">))</span>
            <span class="n">best_concs</span> <span class="o">=</span> <span class="n">grouped_mins</span><span class="p">[</span><span class="n">best_index</span><span class="p">]</span>
            <span class="c"># Keep the best-scoring combination of sequences</span>
            <span class="n">best_combo</span> <span class="o">=</span> <span class="n">combos</span><span class="p">[</span><span class="n">best_index</span><span class="p">]</span>

            <span class="c"># Scoring - mean min free conc, minimum free conc</span>
            <span class="n">mean_score</span> <span class="o">=</span> <span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">best_concs</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">best_concs</span><span class="p">))</span> <span class="o">/</span> <span class="n">conc</span>
            <span class="n">min_score</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">best_concs</span><span class="p">)</span> <span class="o">/</span> <span class="n">conc</span>
            <span class="n">met</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">best_concs</span><span class="p">:</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">x</span> <span class="o">/</span> <span class="n">conc</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_score</span><span class="p">:</span>
                    <span class="n">met</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="c"># Did the minimum score improve? If not, increment counter</span>
            <span class="k">if</span> <span class="n">min_score</span> <span class="o">==</span> <span class="n">last_min_score</span><span class="p">:</span>
                <span class="n">n_attempted</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">n_attempted</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">oligo_list</span> <span class="o">=</span> <span class="n">best_combo</span>

            <span class="c"># Record how much time has passed</span>
            <span class="n">timer</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">time_start</span>

            <span class="c"># Log</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">loop_count</span><span class="p">,</span>
                     <span class="n">n_attempted</span><span class="p">,</span>
                     <span class="n">mean_score</span><span class="p">,</span>
                     <span class="n">min_score</span><span class="p">,</span>
                     <span class="n">met</span><span class="p">,</span>
                     <span class="n">timer</span><span class="p">,</span>
                     <span class="n">oligo_list</span><span class="p">)</span>

            <span class="c"># Script stop decision:</span>
            <span class="c"># Stop looping if either:</span>
            <span class="c">#   1. the threshold is met</span>
            <span class="c">#   2. the minimum score has stopped improving by n_attempt</span>
            <span class="k">if</span> <span class="n">met</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">oligo_list</span><span class="p">):</span>
                <span class="k">break</span>
            <span class="k">if</span> <span class="n">n_attempted</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_attempt</span><span class="p">:</span>
                <span class="k">break</span>

            <span class="n">last_min_score</span> <span class="o">=</span> <span class="n">min_score</span>

        <span class="k">return</span> <span class="n">oligo_list</span>
</div>
<div class="viewcode-block" id="OrthoSeq.log"><a class="viewcode-back" href="../../../pymbt.sequence_generation.html#pymbt.sequence_generation.orthogonal_sequences.OrthoSeq.log">[docs]</a>    <span class="k">def</span> <span class="nf">log</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">loop_count</span><span class="p">,</span>
            <span class="n">n_attempted</span><span class="p">,</span>
            <span class="n">mean_score</span><span class="p">,</span>
            <span class="n">min_score</span><span class="p">,</span>
            <span class="n">met</span><span class="p">,</span>
            <span class="n">timer</span><span class="p">,</span>
            <span class="n">oligo_list</span><span class="p">):</span>
        <span class="sd">&#39;&#39;&#39;provides data trail and allows resuming/passing along</span>
<span class="sd">           an extended attempt&#39;&#39;&#39;</span>

        <span class="c"># Initial file setup for logging</span>
        <span class="k">if</span> <span class="n">loop_count</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="c"># Use date, time, host, and peptide sequence to uniquify file</span>
            <span class="n">current_date</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s">&#39;%Y%m</span><span class="si">%d</span><span class="s">--%H%M%S&#39;</span><span class="p">)</span>
            <span class="n">host</span> <span class="o">=</span> <span class="n">socket</span><span class="o">.</span><span class="n">gethostname</span><span class="p">()</span>
            <span class="n">cur_id</span> <span class="o">=</span> <span class="n">current_date</span> <span class="o">+</span> <span class="s">&#39;-&#39;</span> <span class="o">+</span> <span class="n">host</span> <span class="o">+</span> <span class="s">&#39;-&#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">prot_seq</span>
            <span class="n">cur_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span>
            <span class="n">file_prefix</span> <span class="o">=</span> <span class="s">&#39;</span><span class="si">%s</span><span class="s">/</span><span class="si">%s</span><span class="s">-oligo_calc_&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">cur_dir</span><span class="p">,</span> <span class="n">cur_id</span><span class="p">)</span>

            <span class="c"># Write out setup info file</span>
            <span class="n">info_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_prefix</span> <span class="o">+</span> <span class="s">&#39;info.txt&#39;</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span>
            <span class="n">info_csv</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">info_file</span><span class="p">,</span> <span class="n">quoting</span><span class="o">=</span><span class="n">csv</span><span class="o">.</span><span class="n">QUOTE_MINIMAL</span><span class="p">)</span>
            <span class="n">info_csv</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="s">&#39;sequence&#39;</span><span class="p">,</span> <span class="s">&#39;oligo_n&#39;</span><span class="p">,</span> <span class="s">&#39;oligo_m&#39;</span><span class="p">])</span>
            <span class="n">info_csv</span><span class="o">.</span><span class="n">writerow</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">prot_seq</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">m</span><span class="p">])</span>
            <span class="n">info_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

            <span class="c"># Set up data log file</span>
            <span class="n">data_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_prefix</span> <span class="o">+</span> <span class="s">&#39;data.csv&#39;</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span>
            <span class="n">data_csv</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">data_file</span><span class="p">,</span> <span class="n">quoting</span><span class="o">=</span><span class="n">csv</span><span class="o">.</span><span class="n">QUOTE_MINIMAL</span><span class="p">)</span>
            <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;loop&#39;</span><span class="p">,</span>
                    <span class="s">&#39;n_attempted&#39;</span><span class="p">,</span>
                    <span class="s">&#39;mean_score&#39;</span><span class="p">,</span>
                    <span class="s">&#39;min_score&#39;</span><span class="p">,</span>
                    <span class="s">&#39;met&#39;</span><span class="p">,</span>
                    <span class="s">&#39;time&#39;</span><span class="p">]</span>
            <span class="n">data_csv</span><span class="o">.</span><span class="n">writerow</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span>
            <span class="n">data_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="c"># Write the current set of oligos to file (enables resuming)</span>
        <span class="n">oligo_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_prefix</span> <span class="o">+</span> <span class="s">&#39;latest.txt&#39;</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">oligo_list</span><span class="p">:</span>
            <span class="n">oligo_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
        <span class="n">oligo_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="c"># Update the data log file</span>
        <span class="n">data_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_prefix</span> <span class="o">+</span> <span class="s">&#39;data.csv&#39;</span><span class="p">,</span> <span class="s">&#39;a&#39;</span><span class="p">)</span>
        <span class="n">data_csv</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">data_file</span><span class="p">,</span> <span class="n">quoting</span><span class="o">=</span><span class="n">csv</span><span class="o">.</span><span class="n">QUOTE_MINIMAL</span><span class="p">)</span>
        <span class="n">d</span> <span class="o">=</span> <span class="p">[</span><span class="n">loop_count</span><span class="p">,</span> <span class="n">n_attempted</span><span class="p">,</span> <span class="n">mean_score</span><span class="p">,</span> <span class="n">min_score</span><span class="p">,</span> <span class="n">met</span><span class="p">,</span> <span class="n">timer</span><span class="p">]</span>
        <span class="n">data_csv</span><span class="o">.</span><span class="n">writerow</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
        <span class="n">data_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

</div></div>
<span class="k">def</span> <span class="nf">_n_p_nupack</span><span class="p">(</span><span class="n">sequence_list</span><span class="p">):</span>
    <span class="n">all_concs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sequence_list</span><span class="p">:</span>
        <span class="n">n_np</span> <span class="o">=</span> <span class="n">Nupack</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="s">&#39;dna&#39;</span><span class="p">)</span>
        <span class="n">n_concs</span> <span class="o">=</span> <span class="n">n_np</span><span class="o">.</span><span class="n">concentrations</span><span class="p">(</span><span class="mi">2</span><span class="p">)[</span><span class="s">&#39;concentration&#39;</span><span class="p">]</span>
        <span class="n">n_concs</span> <span class="o">=</span> <span class="n">n_concs</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)]</span>
        <span class="n">all_concs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">n_concs</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">all_concs</span>


<span class="k">def</span> <span class="nf">_monomers_concentration</span><span class="p">(</span><span class="nb">input</span><span class="p">):</span>
    <span class="n">sequence_list</span> <span class="o">=</span> <span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">conc</span> <span class="o">=</span> <span class="nb">input</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

    <span class="c"># Run nupack&#39;s &#39;concentrations&#39;</span>
    <span class="n">np</span> <span class="o">=</span> <span class="n">Nupack</span><span class="p">(</span><span class="n">sequence_list</span><span class="p">,</span> <span class="s">&#39;dna&#39;</span><span class="p">)</span>
    <span class="n">nc</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concentrations</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">conc</span><span class="o">=</span><span class="n">conc</span><span class="p">,</span> <span class="n">mfe</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">concs</span> <span class="o">=</span> <span class="n">nc</span><span class="p">[</span><span class="s">&#39;concentration&#39;</span><span class="p">]</span>
    <span class="n">types</span> <span class="o">=</span> <span class="n">nc</span><span class="p">[</span><span class="s">&#39;types&#39;</span><span class="p">]</span>

    <span class="c"># Isolate the unbound monomer concentrations</span>
    <span class="n">free_conc</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">concs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">types</span><span class="p">)</span> <span class="k">if</span> <span class="nb">sum</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">])</span>
    <span class="n">free_fraction</span> <span class="o">=</span> <span class="n">free_conc</span> <span class="o">/</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">conc</span><span class="p">)</span>
    <span class="n">np</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>  <span class="c"># Delete temp dir</span>

    <span class="k">return</span> <span class="n">free_fraction</span>


<span class="k">def</span> <span class="nf">_gen_oligo</span><span class="p">(</span><span class="n">prot_seq</span><span class="p">,</span> <span class="n">min_score</span><span class="p">,</span> <span class="n">freq_threshold</span><span class="p">,</span> <span class="n">weighted</span><span class="p">,</span> <span class="n">conc</span><span class="p">):</span>
    <span class="c"># Generate oligos until one has a self-self interaction below</span>
    <span class="c"># the threshold.</span>
    <span class="n">threshold_met</span> <span class="o">=</span> <span class="bp">False</span>

    <span class="k">while</span> <span class="ow">not</span> <span class="n">threshold_met</span><span class="p">:</span>
        <span class="c"># Generate a randomly or weighted-randomly</span>
        <span class="c"># coding sequence for the specified peptide</span>
        <span class="c"># Ensure it has an mfe of 0</span>
        <span class="k">if</span> <span class="n">weighted</span><span class="p">:</span>
            <span class="n">choice</span> <span class="o">=</span> <span class="n">weighted_codons</span><span class="o">.</span><span class="n">WeightedCodons</span><span class="p">(</span><span class="n">prot_seq</span><span class="p">,</span>
                                                    <span class="n">frequency_table</span><span class="o">=</span><span class="s">&#39;sc&#39;</span><span class="p">,</span>
                                                    <span class="n">material</span><span class="o">=</span><span class="s">&#39;pep&#39;</span><span class="p">)</span>
            <span class="n">new_oligo</span> <span class="o">=</span> <span class="n">choice</span><span class="o">.</span><span class="n">generate</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ft</span> <span class="o">=</span> <span class="n">freq_threshold</span>
            <span class="n">choice</span> <span class="o">=</span> <span class="n">random_codons</span><span class="o">.</span><span class="n">RandomCodons</span><span class="p">(</span><span class="n">prot_seq</span><span class="p">,</span>
                                                <span class="n">threshold</span><span class="o">=</span><span class="n">ft</span><span class="p">)</span>
            <span class="n">new_oligo</span> <span class="o">=</span> <span class="n">choice</span><span class="o">.</span><span class="n">generate</span><span class="p">()</span>

        <span class="c"># Check oligo for self-self binding and mfe</span>
        <span class="n">oligo_monomer</span> <span class="o">=</span> <span class="n">_monomers_concentration</span><span class="p">(([</span><span class="n">new_oligo</span><span class="p">,</span> <span class="n">new_oligo</span><span class="p">],</span> <span class="n">conc</span><span class="p">))</span>
        <span class="n">oligo_np</span> <span class="o">=</span> <span class="n">Nupack</span><span class="p">([</span><span class="n">new_oligo</span><span class="p">],</span> <span class="s">&#39;dna&#39;</span><span class="p">)</span>
        <span class="n">oligo_mfe</span> <span class="o">=</span> <span class="n">oligo_np</span><span class="o">.</span><span class="n">mfe</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">oligo_np</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">oligo_monomer</span> <span class="o">&gt;=</span> <span class="n">min_score</span> <span class="ow">and</span> <span class="n">oligo_mfe</span> <span class="o">==</span> <span class="mf">0.0</span><span class="p">:</span>
            <span class="n">threshold_met</span> <span class="o">=</span> <span class="bp">True</span>

    <span class="k">return</span> <span class="n">new_oligo</span>


<span class="k">def</span> <span class="nf">_multiprocessing_progress</span><span class="p">(</span><span class="n">mp_iterator</span><span class="p">,</span> <span class="n">total_jobs</span><span class="p">,</span> <span class="n">interval</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
    <span class="n">time_start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="k">while</span> <span class="p">(</span><span class="bp">True</span><span class="p">):</span>
        <span class="n">completed</span> <span class="o">=</span> <span class="n">mp_iterator</span><span class="o">.</span><span class="n">_index</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">completed</span> <span class="o">==</span> <span class="n">total_jobs</span><span class="p">):</span>
            <span class="k">print</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span>
            <span class="k">break</span>
        <span class="k">if</span> <span class="n">completed</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="o">&amp;</span> <span class="n">completed</span> <span class="o">%</span> <span class="mi">20</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="n">interval</span><span class="p">)</span>
            <span class="n">remaining</span> <span class="o">=</span> <span class="n">total_jobs</span> <span class="o">-</span> <span class="n">completed</span>
            <span class="n">time_current</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
            <span class="n">rate_current</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">completed</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">time_current</span> <span class="o">-</span> <span class="n">time_start</span><span class="p">)</span>
            <span class="n">time_remaining</span> <span class="o">=</span> <span class="n">remaining</span> <span class="o">/</span> <span class="n">rate_current</span>
            <span class="n">m</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">time_remaining</span><span class="p">)</span> <span class="o">/</span> <span class="mi">60</span>  <span class="c"># minutes</span>
            <span class="n">s</span> <span class="o">=</span> <span class="n">time_remaining</span> <span class="o">-</span> <span class="mi">60</span> <span class="o">*</span> <span class="n">m</span>  <span class="c"># seconds</span>
            <span class="k">print</span> <span class="s">&quot;Remaining: </span><span class="si">%s</span><span class="s">:</span><span class="si">%.2f</span><span class="s"> min(s)</span><span class="se">\r</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="o">.</span><span class="mi">1</span><span class="p">)</span>


<div class="viewcode-block" id="remove_redundant"><a class="viewcode-back" href="../../../pymbt.sequence_generation.html#pymbt.sequence_generation.orthogonal_sequences.remove_redundant">[docs]</a><span class="k">def</span> <span class="nf">remove_redundant</span><span class="p">(</span><span class="n">seq_list</span><span class="p">,</span> <span class="n">oligo_params</span><span class="p">):</span>
    <span class="k">while</span> <span class="bp">True</span><span class="p">:</span>
        <span class="n">counted</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">(</span><span class="n">seq_list</span><span class="p">)</span>
        <span class="n">redundant</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">new_oligos</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">counted</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">value</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">value</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
                    <span class="n">redundant</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">redundant</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">redundant</span><span class="p">:</span>
                <span class="c"># Remove redundant entry - the one towards the end of the list</span>
                <span class="n">r_index</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_list</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">seq_list</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
                <span class="n">seq_list</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">r_index</span><span class="p">)</span>
                <span class="n">new_oligos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_gen_oligo</span><span class="p">(</span><span class="o">*</span><span class="n">oligo_params</span><span class="p">))</span>
            <span class="n">seq_list</span> <span class="o">+=</span> <span class="n">new_oligos</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">break</span>
    <span class="k">return</span> <span class="n">seq_list</span>

</div>
<div class="viewcode-block" id="unflatten"><a class="viewcode-back" href="../../../pymbt.sequence_generation.html#pymbt.sequence_generation.orthogonal_sequences.unflatten">[docs]</a><span class="k">def</span> <span class="nf">unflatten</span><span class="p">(</span><span class="n">input_list</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
    <span class="n">list_copy</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">input_list</span><span class="p">]</span>
    <span class="n">unflattened</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">list_copy</span><span class="p">)</span> <span class="o">/</span> <span class="n">n</span><span class="p">):</span>
        <span class="n">temp_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
            <span class="n">temp_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">list_copy</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">))</span>
        <span class="n">unflattened</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">temp_list</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">unflattened</span>

</div>
<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">ConfigParser</span>

    <span class="c"># Read config</span>
    <span class="n">current_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">__file__</span><span class="p">))</span>
    <span class="n">config_path</span> <span class="o">=</span> <span class="n">current_path</span> <span class="o">+</span> <span class="s">&#39;/orthogonal_sequences.cfg&#39;</span>
    <span class="n">config</span> <span class="o">=</span> <span class="n">ConfigParser</span><span class="o">.</span><span class="n">RawConfigParser</span><span class="p">()</span>
    <span class="n">config</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">config_path</span><span class="p">)</span>

    <span class="c"># Parse config into parameters</span>
    <span class="n">sequence</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&#39;Main Settings&#39;</span><span class="p">,</span> <span class="s">&#39;protein sequence&#39;</span><span class="p">)</span>
    <span class="n">n</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">getint</span><span class="p">(</span><span class="s">&#39;Main Settings&#39;</span><span class="p">,</span> <span class="s">&#39;n sequences&#39;</span><span class="p">)</span>
    <span class="n">m</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">getint</span><span class="p">(</span><span class="s">&#39;Main Settings&#39;</span><span class="p">,</span> <span class="s">&#39;m sequences&#39;</span><span class="p">)</span>
    <span class="n">T</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">getfloat</span><span class="p">(</span><span class="s">&#39;Main Settings&#39;</span><span class="p">,</span> <span class="s">&#39;temperature&#39;</span><span class="p">)</span>
    <span class="n">min_score</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">getfloat</span><span class="p">(</span><span class="s">&#39;Main Settings&#39;</span><span class="p">,</span> <span class="s">&#39;min unbound monomer score&#39;</span><span class="p">)</span>
    <span class="n">con</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">getfloat</span><span class="p">(</span><span class="s">&#39;Main Settings&#39;</span><span class="p">,</span> <span class="s">&#39;species concentration&#39;</span><span class="p">)</span>
    <span class="n">wholemat</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">getboolean</span><span class="p">(</span><span class="s">&#39;Main Settings&#39;</span><span class="p">,</span> <span class="s">&#39;whole matrix&#39;</span><span class="p">)</span>
    <span class="n">nm</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">getfloat</span><span class="p">(</span><span class="s">&#39;Main Settings&#39;</span><span class="p">,</span> <span class="s">&#39;no increase limit&#39;</span><span class="p">)</span>
    <span class="n">resume</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&#39;Main Settings&#39;</span><span class="p">,</span> <span class="s">&#39;resume&#39;</span><span class="p">)</span>

    <span class="c"># Run OrthoSeq.orthogonal_sequences</span>
    <span class="n">o_seq</span> <span class="o">=</span> <span class="n">OrthoSeq</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">T</span><span class="o">=</span><span class="n">T</span><span class="p">,</span> <span class="n">min_score</span><span class="o">=</span><span class="n">min_score</span><span class="p">,</span> <span class="n">n_attempt</span><span class="o">=</span><span class="n">nm</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">resume</span> <span class="o">==</span> <span class="s">&#39;False&#39;</span><span class="p">:</span>
        <span class="n">o_seq</span><span class="o">.</span><span class="n">start</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">m</span><span class="o">=</span><span class="n">m</span><span class="p">,</span> <span class="n">wholemat</span><span class="o">=</span><span class="n">wholemat</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">resfile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">resume</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span>
        <span class="n">resume</span> <span class="o">=</span> <span class="n">resfile</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
        <span class="n">resfile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="n">resume</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">resume</span><span class="p">]</span>
        <span class="n">o_seq</span><span class="o">.</span><span class="n">start</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">m</span><span class="o">=</span><span class="n">m</span><span class="p">,</span> <span class="n">wholemat</span><span class="o">=</span><span class="n">wholemat</span><span class="p">,</span> <span class="n">resume</span><span class="o">=</span><span class="n">resume</span><span class="p">)</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../../../index.html">pymbt 0.1a documentation</a> &raquo;</li>
          <li><a href="../../index.html" >Module code</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Nick Bolten.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>