

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>oligo_synthesis Package &mdash; pymbt 0.1a documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '0.1a',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="pymbt 0.1a documentation" href="index.html" />
    <link rel="up" title="pymbt Package" href="pymbt.html" />
    <link rel="next" title="sequence_analysis Package" href="pymbt.sequence_analysis.html" />
    <link rel="prev" title="alignment Package" href="pymbt.alignment.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pymbt.sequence_analysis.html" title="sequence_analysis Package"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="pymbt.alignment.html" title="alignment Package"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">pymbt 0.1a documentation</a> &raquo;</li>
          <li><a href="pymbt.html" accesskey="U">pymbt Package</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="oligo-synthesis-package">
<h1>oligo_synthesis Package<a class="headerlink" href="#oligo-synthesis-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-pymbt.oligo_synthesis.gene_splitter">
<span id="gene-splitter-module"></span><h2><tt class="xref py py-mod docutils literal"><span class="pre">gene_splitter</span></tt> Module<a class="headerlink" href="#module-pymbt.oligo_synthesis.gene_splitter" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="pymbt.oligo_synthesis.gene_splitter.GeneSplitter">
<em class="property">class </em><tt class="descclassname">pymbt.oligo_synthesis.gene_splitter.</tt><tt class="descname">GeneSplitter</tt><big>(</big><em>seq</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pymbt/oligo_synthesis/gene_splitter.html#GeneSplitter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pymbt.oligo_synthesis.gene_splitter.GeneSplitter" title="Permalink to this definition">¶</a></dt>
<dd><p>Split a large (~&lt;10kb) sequence into smaller ones that are easier to
clone, either via oligo assembly or PCR. Store and write the results.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>seq</strong> (<em>str.</em>) &#8211; Input sequence.</li>
<li><strong>**kwargs</strong> &#8211; <p>Keyword arguments to pass to split_gene.</p>
</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="pymbt.oligo_synthesis.gene_splitter.GeneSplitter.write">
<tt class="descname">write</tt><big>(</big><em>path</em><big>)</big><a class="reference internal" href="_modules/pymbt/oligo_synthesis/gene_splitter.html#GeneSplitter.write"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pymbt.oligo_synthesis.gene_splitter.GeneSplitter.write" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="function">
<dt id="pymbt.oligo_synthesis.gene_splitter.find_best">
<tt class="descclassname">pymbt.oligo_synthesis.gene_splitter.</tt><tt class="descname">find_best</tt><big>(</big><em>walked</em>, <em>max_len</em>, <em>seq_len</em>, <em>force_exhaustive=False</em><big>)</big><a class="reference internal" href="_modules/pymbt/oligo_synthesis/gene_splitter.html#find_best"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pymbt.oligo_synthesis.gene_splitter.find_best" title="Permalink to this definition">¶</a></dt>
<dd><p>Search for the optimal overlap combination.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>walked</strong> &#8211; list of lists of the output to context_walk, representing</td>
</tr>
</tbody>
</table>
<p>a collection of each potential overlap&#8217;s evaluated windows.
:type walked: list.
:param max_len: Maximum length of each split-up output sequence.
:type max_len: int.
:param force_exhaustive: Forces algorithm to find global optimum
exhaustively. Can dramatically slow down the computation.
:type force_exhaustive: bool.</p>
</dd></dl>

<dl class="function">
<dt id="pymbt.oligo_synthesis.gene_splitter.remove_from_left">
<tt class="descclassname">pymbt.oligo_synthesis.gene_splitter.</tt><tt class="descname">remove_from_left</tt><big>(</big><em>pre_combo_list</em>, <em>max_len</em><big>)</big><a class="reference internal" href="_modules/pymbt/oligo_synthesis/gene_splitter.html#remove_from_left"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pymbt.oligo_synthesis.gene_splitter.remove_from_left" title="Permalink to this definition">¶</a></dt>
<dd><p>Traverse list of lists of (start, end, score) tuples of the type
returned by context_walk. Removes any entries that cannot bridge the
gap to the next set of overlaps (distance greater than max_distance).
This greatly speeds up computation time by removing impossible combinations
before they are evaluated computationally.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>pre_combo_list</strong> &#8211; list of lists of the output to context_walk,</td>
</tr>
</tbody>
</table>
<p>representing a collection of each potential overlap&#8217;s evaluated windows.
:type pre_combo_list: list.
:param max_len: Maximum length of each split-up output sequence.
:type max_len: int.</p>
</dd></dl>

<dl class="function">
<dt id="pymbt.oligo_synthesis.gene_splitter.remove_from_right">
<tt class="descclassname">pymbt.oligo_synthesis.gene_splitter.</tt><tt class="descname">remove_from_right</tt><big>(</big><em>pre_combo_list</em>, <em>max_len</em><big>)</big><a class="reference internal" href="_modules/pymbt/oligo_synthesis/gene_splitter.html#remove_from_right"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pymbt.oligo_synthesis.gene_splitter.remove_from_right" title="Permalink to this definition">¶</a></dt>
<dd><p>Does the same thing as remove_from_left but in the reverse direction.
These two functions have to be paired repeatedly in order to fully trim
the sequence (remove_nonspanning will do this)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>pre_combo_list</strong> &#8211; list of lists of the output to context_walk,</td>
</tr>
</tbody>
</table>
<p>representing a collection of each potential overlap&#8217;s evaluated windows.
:type pre_combo_list: list.
:param max_len: Maximum length of each split-up output sequence.
:type max_len: int.</p>
</dd></dl>

<dl class="function">
<dt id="pymbt.oligo_synthesis.gene_splitter.remove_nonspanning">
<tt class="descclassname">pymbt.oligo_synthesis.gene_splitter.</tt><tt class="descname">remove_nonspanning</tt><big>(</big><em>pre_combo_list</em>, <em>max_len</em><big>)</big><a class="reference internal" href="_modules/pymbt/oligo_synthesis/gene_splitter.html#remove_nonspanning"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pymbt.oligo_synthesis.gene_splitter.remove_nonspanning" title="Permalink to this definition">¶</a></dt>
<dd><p>Trims list of lists of (start, end, score) tuples to those
that actually have a chance of spanning the sequence</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>pre_combo_list</strong> &#8211; list of lists of the output to context_walk,</td>
</tr>
</tbody>
</table>
<p>representing a collection of each potential overlap&#8217;s evaluated windows.
:type pre_combo_list: list.
:param max_len: Maximum length of each split-up output sequence.
:type max_len: int.</p>
</dd></dl>

<dl class="function">
<dt id="pymbt.oligo_synthesis.gene_splitter.split_gene">
<tt class="descclassname">pymbt.oligo_synthesis.gene_splitter.</tt><tt class="descname">split_gene</tt><big>(</big><em>seq</em>, <em>max_len=1100</em>, <em>min_context=200</em>, <em>core=60</em>, <em>context=90</em>, <em>step=10</em>, <em>force_exhaustive=False</em><big>)</big><a class="reference internal" href="_modules/pymbt/oligo_synthesis/gene_splitter.html#split_gene"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pymbt.oligo_synthesis.gene_splitter.split_gene" title="Permalink to this definition">¶</a></dt>
<dd><p>Split a large (~&lt;10kb) sequence into smaller ones that are easier to clone,
either via oligo assembly or PCR.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>max_len</strong> (<em>int.</em>) &#8211; Maximum length of each split-up output sequence.</li>
<li><strong>min_context</strong> &#8211; Sets minimum size of each potential overlap</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>region to evaluate.
:type min_context: int.
:param core: Window size.
:type core: int.
:param context: Amount (in bp) of context to evaluate on either side
of the window.
:type context: int.
:param step: Step size for the windows.
:type step: int.
:param force_exhaustive: Forces algorithm to find global optimum
exhaustively. Can dramatically slow down the computation.
:type force_exhaustive: bool.</p>
</dd></dl>

</div>
<div class="section" id="module-pymbt.oligo_synthesis.oligo_assembly">
<span id="oligo-assembly-module"></span><h2><tt class="xref py py-mod docutils literal"><span class="pre">oligo_assembly</span></tt> Module<a class="headerlink" href="#module-pymbt.oligo_synthesis.oligo_assembly" title="Permalink to this headline">¶</a></h2>
<p>Module supplying methods/classes for generating
overlapping oligo sequences from a gene sequence.</p>
<dl class="class">
<dt id="pymbt.oligo_synthesis.oligo_assembly.OligoAssembly">
<em class="property">class </em><tt class="descclassname">pymbt.oligo_synthesis.oligo_assembly.</tt><tt class="descname">OligoAssembly</tt><big>(</big><em>seq</em>, <em>primers=False</em>, <em>primer_tm=60</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pymbt/oligo_synthesis/oligo_assembly.html#OligoAssembly"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pymbt.oligo_synthesis.oligo_assembly.OligoAssembly" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">object</span></tt></p>
<p>Split a sequence into overlapping oligonucleotides with equal-Tm
overlaps. Contains results and output methods.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>seq</strong> (<em>str.</em>) &#8211; Input sequence (DNA).</li>
<li><strong>primers</strong> &#8211; Design cloning primers that bind the termini of the</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>input sequence.
:type primers: bool.
:param primer_tm: Ideal Tm for the cloning primers, if applicable.
:type primer_tm: float.
:param <a href="#id3"><span class="problematic" id="id4">**</span></a>kwargs: Keyword arguments to pass to oligo_calc.</p>
<dl class="method">
<dt id="pymbt.oligo_synthesis.oligo_assembly.OligoAssembly.write">
<tt class="descname">write</tt><big>(</big><em>outpath</em><big>)</big><a class="reference internal" href="_modules/pymbt/oligo_synthesis/oligo_assembly.html#OligoAssembly.write"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pymbt.oligo_synthesis.oligo_assembly.OligoAssembly.write" title="Permalink to this definition">¶</a></dt>
<dd><p>Write results out to a csv (comma-separated) file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>outpath</strong> (<em>str.</em>) &#8211; path to csv file, including .csv extension.</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="pymbt.oligo_synthesis.oligo_assembly.oligo_calc">
<tt class="descclassname">pymbt.oligo_synthesis.oligo_assembly.</tt><tt class="descname">oligo_calc</tt><big>(</big><em>seq</em>, <em>tm=72</em>, <em>oligo_size=120</em>, <em>require_even=True</em>, <em>start_5=True</em><big>)</big><a class="reference internal" href="_modules/pymbt/oligo_synthesis/oligo_assembly.html#oligo_calc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pymbt.oligo_synthesis.oligo_assembly.oligo_calc" title="Permalink to this definition">¶</a></dt>
<dd><p>Split a sequence into overlapping oligonucleotides with equal-Tm
overlaps.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>seq</strong> (<em>str.</em>) &#8211; Input sequence (DNA).</li>
<li><strong>tm</strong> (<em>float.</em>) &#8211; Ideal Tm of the overlaps, in degrees C.</li>
<li><strong>oligo_size</strong> (<em>int.</em>) &#8211; Maximum oligo size (e.g. price-point jump size).</li>
<li><strong>require_even</strong> (<em>bool.</em>) &#8211; Require that the number of oligonucleotides is even.</li>
<li><strong>start_5</strong> (<em>bool.</em>) &#8211; Require that the first oligo&#8217;s terminal side is 5&#8217;.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-pymbt.oligo_synthesis.structure_windows">
<span id="structure-windows-module"></span><h2><tt class="xref py py-mod docutils literal"><span class="pre">structure_windows</span></tt> Module<a class="headerlink" href="#module-pymbt.oligo_synthesis.structure_windows" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pymbt.oligo_synthesis.structure_windows.context_walk">
<tt class="descclassname">pymbt.oligo_synthesis.structure_windows.</tt><tt class="descname">context_walk</tt><big>(</big><em>seq</em>, <em>core_len</em>, <em>context_len</em>, <em>step</em><big>)</big><a class="reference internal" href="_modules/pymbt/oligo_synthesis/structure_windows.html#context_walk"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pymbt.oligo_synthesis.structure_windows.context_walk" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate context-dependent &#8216;non-boundedness&#8217; series of scores for a given
DNA sequence. Uses NUPACK&#8217;s pair probabilities to derive the score.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>seq</strong> (<em>str.</em>) &#8211; Input sequence.</li>
<li><strong>core_len</strong> (<em>int.</em>) &#8211; Window size in base pairs.</li>
<li><strong>context_len</strong> &#8211; The number of bases of context to use when analyzing</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>each window.
:type context_len: int.
:param step: The number of base pairs to move for each new window.
:type step: int.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">oligo_synthesis Package</a><ul>
<li><a class="reference internal" href="#module-pymbt.oligo_synthesis.gene_splitter"><tt class="docutils literal"><span class="pre">gene_splitter</span></tt> Module</a></li>
<li><a class="reference internal" href="#module-pymbt.oligo_synthesis.oligo_assembly"><tt class="docutils literal"><span class="pre">oligo_assembly</span></tt> Module</a></li>
<li><a class="reference internal" href="#module-pymbt.oligo_synthesis.structure_windows"><tt class="docutils literal"><span class="pre">structure_windows</span></tt> Module</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="pymbt.alignment.html"
                        title="previous chapter">alignment Package</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="pymbt.sequence_analysis.html"
                        title="next chapter">sequence_analysis Package</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/pymbt.oligo_synthesis.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pymbt.sequence_analysis.html" title="sequence_analysis Package"
             >next</a> |</li>
        <li class="right" >
          <a href="pymbt.alignment.html" title="alignment Package"
             >previous</a> |</li>
        <li><a href="index.html">pymbt 0.1a documentation</a> &raquo;</li>
          <li><a href="pymbt.html" >pymbt Package</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Nick Bolten.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>